### PR TITLE
Add SMALLINT and TINYINT to Presto

### LINKED ISSUE #4513: Add SMALLINT, TINYINT, INT types to Presto

### PR DESCRIPTION
**Pull request by raghavsethi**
Closes #4513.

### CODE REVIEW

**Comment by cberner** [comment]
Looks like the tests are failing. Can you assign back to me when you've fixed that? Also, did you run this in the verifier?

**Comment by raghavsethi** [comment]
Verifier run was clean.

**Comment by raghavsethi** [comment]
@cberner fixed the test and ran the verifier, back to you.

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java [hunk 1]
```diff
@@ -38,6 +38,22 @@ private MathFunctions() {}
 
     @Description("absolute value")
     @ScalarFunction("abs")
+    @SqlType(StandardTypes.TINYINT)
+    public static long absTinyInt(@SqlType(StandardTypes.TINYINT) long num)
```
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java]: style note: the casing here isn't consistent with other places. It looks like we use `bigint` as a single word in other places, and you used `Smallint` in `SmallintType`, so I would stick with casing it as a single word

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java [hunk 2]
```diff
@@ -38,6 +38,22 @@ private MathFunctions() {}
...
+    {
+        return Math.abs(num);
+    }
+
+    @Description("absolute value")
+    @ScalarFunction("abs")
+    @SqlType(StandardTypes.SMALLINT)
+    public static long absSmallInt(@SqlType(StandardTypes.SMALLINT) long num)
```
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java]: same for all of these

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java [hunk 3]
```diff
@@ -38,6 +38,22 @@ private MathFunctions() {}
 
     @Description("absolute value")
     @ScalarFunction("abs")
+    @SqlType(StandardTypes.TINYINT)
+    public static long absTinyInt(@SqlType(StandardTypes.TINYINT) long num)
+    {
+        return Math.abs(num);
```
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java]: This is going to fail for `Byte.MIN_VALUE` as it will become `Byte.MAX_VALUE + 1`. Also, it looks like the implementation for `Integer` is broken as well

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java [hunk 4]
```diff
@@ -38,6 +38,22 @@ private MathFunctions() {}
...
+    }
+
+    @Description("absolute value")
+    @ScalarFunction("abs")
+    @SqlType(StandardTypes.SMALLINT)
+    public static long absSmallInt(@SqlType(StandardTypes.SMALLINT) long num)
+    {
+        return Math.abs(num);
```
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java]: same issue with min value

File: presto-main/src/main/java/com/facebook/presto/type/DecimalCasts.java [hunk 5]
```diff
@@ -238,6 +244,114 @@ public static Slice integerToLongDecimal(long value, long precision, long scale,
...
+    public static long shortDecimalToSmallint(long decimal, long precision, long scale, long tenToScale)
+    {
+        // this rounds the decimal value to the nearest integral value
+        long longResult = (decimal + tenToScale / 2) / tenToScale;
+        if (decimal < 0) {
+            longResult = -((-decimal + tenToScale / 2) / tenToScale);
+        }
+
+        short shortValue = (short) longResult;
+        if (shortValue != longResult) {
```
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/type/DecimalCasts.java]: use `Shorts.checkedCast` like you did for bigint -> smallint?

File: presto-main/src/main/java/com/facebook/presto/type/DecimalCasts.java [hunk 6]
```diff
@@ -238,6 +244,114 @@ public static Slice integerToLongDecimal(long value, long precision, long scale,
...
+    public static long shortDecimalToTinyint(long decimal, long precision, long scale, long tenToScale)
+    {
+        // this rounds the decimal value to the nearest integral value
+        long longResult = (decimal + tenToScale / 2) / tenToScale;
+        if (decimal < 0) {
+            longResult = -((-decimal + tenToScale / 2) / tenToScale);
+        }
+
+        byte shortValue = (byte) longResult;
+        if (shortValue != longResult) {
```
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/type/DecimalCasts.java]: use `SignedByte`?

File: presto-main/src/main/java/com/facebook/presto/type/DecimalCasts.java [hunk 7]
```diff
@@ -238,6 +244,114 @@ public static Slice integerToLongDecimal(long value, long precision, long scale,
...
+    @UsedByGeneratedCode
+    public static long shortDecimalToTinyint(long decimal, long precision, long scale, long tenToScale)
+    {
+        // this rounds the decimal value to the nearest integral value
+        long longResult = (decimal + tenToScale / 2) / tenToScale;
+        if (decimal < 0) {
+            longResult = -((-decimal + tenToScale / 2) / tenToScale);
+        }
+
+        byte shortValue = (byte) longResult;
```
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/type/DecimalCasts.java]: should probably be `tinyValue`

File: presto-main/src/main/java/com/facebook/presto/type/SmallintOperators.java [hunk 8]
```diff
@@ -0,0 +1,223 @@
...
+    private SmallintOperators()
+    {
+    }
+
+    @ScalarOperator(ADD)
+    @SqlType(StandardTypes.SMALLINT)
+    public static long add(@SqlType(StandardTypes.SMALLINT) long left, @SqlType(StandardTypes.SMALLINT) long right)
+    {
+        try {
+            return Math.addExact((short) left, (short) right);
```
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/type/SmallintOperators.java]: I don't think this works. `Math.addExact` throws an exception if the result overflows an `int`, but you're trying to operate on `short`

File: presto-main/src/main/java/com/facebook/presto/type/SmallintOperators.java [hunk 9]
```diff
@@ -0,0 +1,223 @@
...
+            throw new PrestoException(NUMERIC_VALUE_OUT_OF_RANGE, e);
+        }
+    }
+
+    @ScalarOperator(SUBTRACT)
+    @SqlType(StandardTypes.SMALLINT)
+    public static long subtract(@SqlType(StandardTypes.SMALLINT) long left, @SqlType(StandardTypes.SMALLINT) long right)
+    {
+        try {
+            return Math.subtractExact((short) left, (short) right);
```
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/type/SmallintOperators.java]: Same comment for all of these

File: presto-main/src/main/java/com/facebook/presto/type/TinyintOperators.java [hunk 10]
```diff
@@ -0,0 +1,218 @@
...
+    private TinyintOperators()
+    {
+    }
+
+    @ScalarOperator(ADD)
+    @SqlType(StandardTypes.TINYINT)
+    public static long add(@SqlType(StandardTypes.TINYINT) long left, @SqlType(StandardTypes.TINYINT) long right)
+    {
+        try {
+            return Math.addExact((byte) left, (byte) right);
```
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/type/TinyintOperators.java]: Same comment about overflow

File: presto-spi/src/main/java/com/facebook/presto/spi/type/SmallintType.java [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+    public void writeLong(BlockBuilder blockBuilder, long value)
+    {
+        if (value > Short.MAX_VALUE) {
+            throw new PrestoException(NUMERIC_VALUE_OUT_OF_RANGE, String.format("Value %d exceeds MAX_SHORT", value));
+        }
+        else if (value < Short.MIN_VALUE) {
+            throw new PrestoException(NUMERIC_VALUE_OUT_OF_RANGE, String.format("Value %d is less than MIN_SHORT", value));
+        }
+
+        blockBuilder.writeShort((int) value).closeEntry();
```
Comment (cberner) [inline presto-spi/src/main/java/com/facebook/presto/spi/type/SmallintType.java]: casting as an `int` does not look correct

**Comment by cberner** [comment]
The first comment looks like it has a lot of issues with overflow handling. Also, rather than copying the tests for another type, I would focus on testing things that are likely to be different for these types (like overflow).

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java [hunk 12]
```diff
@@ -1389,8 +1389,8 @@ else if (firstDependencies.stream().allMatch(right.canResolvePredicate()) && sec
 
     private void addCoercionForJoinCriteria(Join node, Expression leftExpression, Expression rightExpression)
     {
-        Type leftType = analysis.getType(leftExpression);
-        Type rightType = analysis.getType(rightExpression);
+        Type leftType = analysis.getTypeWithCoercions(leftExpression);
```
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java]: Explain in the comment message what was wrong

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveUtil.java [hunk 13]
```diff
@@ -387,6 +391,26 @@ public static NullableValue parsePartitionValue(String partitionName, String val
...
         }
 
+        if (TINYINT.equals(type)) {
+            if (isNull) {
+                return NullableValue.asNull(TINYINT);
+            }
+            if (value.isEmpty()) {
+                return NullableValue.of(TINYINT, 0L);
+            }
+            return NullableValue.of(TINYINT, parseLong(value));
```
Comment (cberner) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveUtil.java]: Is `parseLong` really the right thing to do here?

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveUtil.java [hunk 14]
```diff
@@ -387,6 +391,26 @@ public static NullableValue parsePartitionValue(String partitionName, String val
...
+        }
+
+        if (SMALLINT.equals(type)) {
+            if (isNull) {
+                return NullableValue.asNull(SMALLINT);
+            }
+            if (value.isEmpty()) {
+                return NullableValue.of(SMALLINT, 0L);
+            }
+            return NullableValue.of(SMALLINT, parseLong(value));
```
Comment (cberner) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveUtil.java]: same comment

File: presto-hive/src/main/java/com/facebook/presto/hive/parquet/ParquetPageSourceFactory.java [hunk 15]
```diff
@@ -229,9 +232,22 @@ public static ParquetPageSource createParquetPageSource(
     // TODO: support complex types
     private static boolean columnTypeSupported(List<HiveColumnHandle> columns)
     {
-        return columns.stream()
+        if (SUPPORTED_COLUMN_TYPES.contains(columns.get(0))) {
+            System.out.println("lol");
```
Comment (cberner) [inline presto-hive/src/main/java/com/facebook/presto/hive/parquet/ParquetPageSourceFactory.java]: remove this :)

**Comment by cberner** [comment]
Back to you for comments

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveUtil.java (hunk 13 above)
Comment (raghavsethi) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveUtil.java]: Oops this seems to be a rebase conflict. It used to be the right thing to do!

File: presto-main/src/main/java/com/facebook/presto/type/SmallintOperators.java (hunk 9 above)
Comment (raghavsethi) [inline presto-main/src/main/java/com/facebook/presto/type/SmallintOperators.java]: I agree that my implementation is broken, but what's really weird is that I have tests that exercise all these code paths.

Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/type/SmallintOperators.java]: Oh ya, that's strange. What might be happening here is that the exception is being thrown from `SmallintType.writeLong` in the test. I would actually change that to be an internal error, since that indicates data corruption in Presto or the connector, as that value could never exist given a proper implementation of all the functions/connectors

Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/type/SmallintOperators.java]: Probably good to change `IntegerType` too

Comment (raghavsethi) [inline presto-main/src/main/java/com/facebook/presto/type/SmallintOperators.java]: Ah the tests fail at the writeLong phase, not when the math is done.

Comment (raghavsethi) [inline presto-main/src/main/java/com/facebook/presto/type/SmallintOperators.java]: Lol sorry didn't reload the page while debugging :)

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveFileFormats.java [hunk 16]
```diff
@@ -248,9 +248,14 @@ public void testParquetUseColumnNames(int rowCount)
     {
         // Write of complex hive data to Parquet is broken
         // TODO: empty arrays or maps with null keys don't seem to work
+        // Parquet does not support DATE
```
Comment (raghavsethi) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHiveFileFormats.java]: I had to add the DATE check for completeness. Should I make it a separate commit or is it OK?

**Comment by raghavsethi** [comment]
Back to you. I added a bunch of smaller cleanup commits for stuff I found while looking at your comments.

File: presto-main/src/main/java/com/facebook/presto/type/TypeRegistry.java [hunk 17]
```diff
@@ -346,6 +350,32 @@ public void addParametricType(ParametricType parametricType)
                         return Optional.empty();
                 }
             }
+            case StandardTypes.TINYINT: {
+                switch (resultTypeBase) {
+                    case StandardTypes.SMALLINT:
+                        return Optional.of(INTEGER);
```
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/type/TypeRegistry.java]: This looks wrong. Shouldn't the common type be `smallint`?

File: presto-main/src/main/java/com/facebook/presto/type/VarcharOperators.java [hunk 18]
```diff
@@ -176,6 +176,32 @@ public static long castToInteger(@SqlType("varchar(x)") Slice slice)
 
     @LiteralParameters("x")
     @ScalarOperator(CAST)
+    @SqlType(StandardTypes.SMALLINT)
+    public static long castToSmallInt(@SqlType("varchar(x)") Slice slice)
```
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/type/VarcharOperators.java]: casing, the `I` in `int` should be lower case

File: presto-main/src/main/java/com/facebook/presto/type/VarcharOperators.java [hunk 19]
```diff
@@ -176,6 +176,32 @@ public static long castToInteger(@SqlType("varchar(x)") Slice slice)
...
+        }
+        catch (Exception e) {
+            throw new PrestoException(INVALID_CAST_ARGUMENT, format("Can not cast '%s' to SMALLINT", slice.toStringUtf8()));
+        }
+    }
+
+    @LiteralParameters("x")
+    @ScalarOperator(CAST)
+    @SqlType(StandardTypes.TINYINT)
+    public static long castToTinyInt(@SqlType("varchar(x)") Slice slice)
```
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/type/VarcharOperators.java]: same

Comment (raghavsethi) [inline presto-main/src/main/java/com/facebook/presto/type/VarcharOperators.java]: Sorry, my find and replace missed a case. I've got them all now.

**Comment by raghavsethi** [comment]
Please refresh the page before you continue reviewing.

**Comment by cberner** [comment]
See comment about type coercion, otherwise looks good. Make sure to run all our internal integration tests before merging though.
